### ISSUE TITLE
Consider distributing ESMPy via PyPI

### ISSUE THREAD

**Issue by billsacks** [esmf-org/esmf#256]
We have had a number of requests to distribute ESMPy via PyPI so that it can be available via pip. e.g., see discussion in https://github.com/conda-forge/esmpy-feedstock/issues/72

Ideally this will be set up using an automated workflow so that it will automatically be updated on PyPI as new ESMF versions come out.

**Comment by ciaransweet** [comment]
This would be really useful! I don't use Conda and it would be a big move to swap to using it, just for the sake of being able to install ESMF without having to do it from source

**Comment by semvijverberg** [comment]
Yes, that would be great! Now it we can't use `hatch` to install our environment because it has to rely on Conda. Maybe there is a workout though?

**Comment by Zeitsperre** [comment]
Would also love to see this. It would simplify the maintenance for more than a few of the projects I work on!

**Comment by anntsay** [comment]
follow up with Ryan to see if he has experience on this and maybe adding to his scope.

**Comment by nannau** [comment]
I too would love to see PyPI support for ESMpy! It would greatly simplify maintenance and internal tooling at my workplace, especially for platforms that don't support Conda. This would also enable us to use tools like xESMF more broadly -- especially [given recent directions](https://www.anaconda.com/pricing). I'd like to help keep this issue active and contribute if I can!

As mentioned, some of this is partially covered in [this issue](https://github.com/conda-forge/esmpy-feedstock/issues/72), but here seems like a more logical place to continue. Please let me know if not!

---

To explore the feasibility of adding PyPI support, I tried a preliminary exercise using the following setup:

System: Ubuntu 22.04.5
Python: 3.12.6

1. Built ESMF locally from this repository.
2. Symlinked the `esmf/lib` directory into` esmf/src/addon/esmpy/src/esmpy` (admittedly, this might be an odd approach—there's likely a cleaner method).
3. Updated the pyproject.toml to include the following under [tool.setuptools.package-data]:
```toml
[tool.setuptools.package-data]
esmpy = [ "lib/libO/Linux.gfortran.64.mpiuni.default/*" ]
```
4. Installed esmpy from the `pyproject.toml` configuration into a fresh Python environment. The binaries were correctly included under `venv/lib/python3.12/site-packages/esmpy/lib/libO/Linux.gfortran.64.mpiuni.default/`.
5. Manually set `$ESMFMKFILE`, which also seemed to work as intended.
6. I built a wheel and source distribution following [these steps ](https://packaging.python.org/en/latest/tutorials/packaging-projects/) and it too seems to have worked well.

I stopped here because I'm unclear about the following:

* How does the Conda installation determine where to locate the ESMF binaries?
* Is the above approach similar to how the project is packaged for Conda?
* Did I miss any critical steps? I did not run comprehensive testing here.

Assuming I'm on the right track and the above steps are generalizable, I understand the tricky part is trying to automate this. There was [some discussion](https://github.com/conda-forge/esmpy-feedstock/issues/72#issuecomment-2344341950) about potentially using docker images from here, but I don't quite have the experience with the esmf CI/CD pipeline to continue without a few hints. Cheers!

**Comment by anntsay** [comment]
Thanks for the comments.  @nannau thanks for letting us know what you have tried so far.  we will take a look at this and add this to the backlog.

**Comment by xylar** [comment]
@billsacks, @anntsay and others, what's the status of PyPI support on your side?  I'm starting to play with it in:
https://github.com/xylar/esmpy-wheels
and
https://github.com/xylar/esmf/tree/add-pypi-support
but I'm not committed to this approach if there's another branch I should be working from instead.

Has this work lost steam, as I had guessed it might have?  Or is there a lot of behind-the-scenes work that I should know about before I duplicate efforts or step on toes?

**Comment by xylar** [comment]
Following up on my previous comment, I've been prototyping a way to distribute self-contained ESMPy wheels, and I'd like to propose a concrete path forward and offer to do the work.

**What I have so far**

A working prototype pipeline at https://github.com/xylar/esmpy-wheels builds fully self-contained, `pip install`-able ESMPy wheels: it builds ESMF + HDF5 + NetCDF-C/Fortran from source, grafts `libesmf_fullylinked` and `esmf.mk` into the pure-Python ESMPy package, and vendors the native dependencies with auditwheel (Linux) / delocate (macOS). Because ESMPy has no compiled extension of its own, the result is one `py3-none-<platform>` wheel per (OS, arch). CI is currently green for serial (mpiuni) + NetCDF on Linux (manylinux_2_28 x86_64) and macOS (arm64), with MPI (MPICH) and additional platforms staged behind an (os, arch, comm) build matrix. Nothing is published to PyPI yet.

This depends on a small set of loader changes in ESMF itself — making `esmf.mk` / library discovery wheel-aware and relocatable, preloading `libmpi` for MPI builds, and resolving variant distribution names such as `esmpy-mpich`. All of these are no-ops for conda, source, and HPC installs.

**Suggested path forward**

1. **ESMF PR — non-Windows loader changes.** A PR with the Linux/macOS + MPI loader changes described above. These are the prerequisite for the wheel build and are backward-compatible with all existing install methods. (Open as #571.)
2. **ESMF PR — Windows changes (follow-up).** A larger follow-up PR (#<WINDOWS-PR>), building on #571, with the changes needed to build ESMF and its dependencies on Windows (MinGW/MSYS2 toolchain, LLP64 fixes, symbol exports, and related fixes), together with the already-staged Windows branch in the loader. (Draft as #572)
3. **Wheel-building repo under esmf-org.** Create a repository on esmf-org (by copying or forking the prototype) to host the wheel-building pipeline and its GitHub Actions CI. Keeping it separate from the main ESMF repo keeps the heavy from-source dependency builds and release automation out of the core repo.
4. **PyPI setup.** Create/curate the project on PyPI and on test.pypi.org (including any variant names such as `esmpy-mpich`), decide on ownership (an esmf-org account and/or maintainer accounts), and configure trusted publishing (OIDC from the wheel repo's Actions) so releases don't depend on long-lived tokens. We can validate the full upload/install round-trip on TestPyPI first.
5. **First release + rollout.** Agree on the initial support matrix (OS / arch / Python / MPI), align wheel versions with ESMF release tags, publish an initial release to TestPyPI and then PyPI, and expand the matrix incrementally.

**Other things to decide along the way**

- Distribution naming and how the serial vs MPI variants relate (`esmpy` vs `esmpy-mpich`), and how `mpi4py` / MPI ABI compatibility is communicated to users.
- Coordination with the conda-forge feedstock maintainers (conda-forge/esmpy-feedstock#72) so the two distribution channels stay consistent.
- Release triggering: whether wheels are built automatically on ESMF version tags or on demand.
- Documentation updates pointing users at `pip install esmpy`.

I'm happy to lead this — open the two ESMF PRs (the non-Windows changes are up as #571 and the Windows changes as #572), contribute the wheel-building repo, and help set up the PyPI / TestPyPI side — following whatever review and ownership process the ESMF team prefers. Feedback on the approach would be very welcome, especially on the esmf-org repo and the PyPI account/ownership decisions, which are yours to make.
